include/verilated_random.h

```diff
@@ -0,0 +1,128 @@
+///
+//*************************************************************************
+#ifndef VERILATOR_VERILATED_RANDOM_H_
+#define VERILATOR_VERILATED_RANDOM_H_
+
+#include "verilated.h"
+
+//=============================================================================
+// VlRandomExpr and subclasses represent expressions for the constraint solver.
+
+class VlRandomExpr VL_NOT_FINAL {
+public:
+    virtual void emit(std::ostream& s) const = 0;
+};
+class VlRandomVar VL_NOT_FINAL : public VlRandomExpr {
+    const char* const m_name;
+    void* const m_ref;
+    const int m_width;
+
+public:
+    VlRandomVar(const char* name, int width, void* ref)
+        : m_name{name}
+        , m_ref{ref}
+        , m_width{width} {}
+    const char* name() const { return m_name; }
+    int width() const { return m_width; }
+    void* ref() const { return m_ref; }
+    virtual void set(unsigned long long) const = 0;
+    void emit(std::ostream& s) const;
+};
+template <typename T>
+class VlRandomVarR final : public VlRandomVar {
+public:
+    VlRandomVarR(const char* name, int width, T* ref)
+        : VlRandomVar{name, width, ref} {}
+    virtual void set(unsigned long long val) const { *(T*)ref() = val; }
+};
+
+class VlRandomConst final : public VlRandomExpr {
+    const unsigned long long m_val;
+    const int m_width;
+
+public:
+    VlRandomConst(unsigned long long val, int width)
+        : m_val{val}
+        , m_width{width} {}
+    void emit(std::ostream& s) const;
+};
+
+class VlRandomExtract final : public VlRandomExpr {
+    const std::shared_ptr<const VlRandomExpr> m_expr;
+    const unsigned m_idx;
+
+public:
+    VlRandomExtract(std::shared_ptr<const VlRandomExpr> expr, unsigned idx)
+        : m_expr{expr}
+        , m_idx{idx} {}
+    void emit(std::ostream& s) const;
+};
+
+class VlRandomBinOp final : public VlRandomExpr {
+    const char* m_op;
+    const std::shared_ptr<const VlRandomExpr> m_lhs, m_rhs;
+
+public:
+    VlRandomBinOp(const char* op, std::shared_ptr<const VlRandomExpr> lhs,
+                  std::shared_ptr<const VlRandomExpr> rhs)
+        : m_op{op}
+        , m_lhs{lhs}
+        , m_rhs{rhs} {}
+    void emit(std::ostream& s) const;
+};
+
+//=============================================================================
+// VlRandomizer is the object holding constraints and variable references.
+
+class VlRandomizer final {
+    // MEMBERS
+    std::vector<std::string> m_constraints;
+    std::map<std::string, std::shared_ptr<const VlRandomVar>> m_vars;
+    char* m_line = nullptr;
+    size_t m_capacity;
+
+    // PRIVATE METHODS
+    std::shared_ptr<const VlRandomExpr> random_constraint(VlRNG& rngr, int bits);
+    int parse_solution(FILE* file);
+
+public:
+    // METHODS
+    // Finds the next solution satisfying the constraints
+    bool next(VlRNG& rngr);
+    template <typename T>
+    void write_var(T& var, int width, const char* name) {
+        auto it = m_vars.find(name);
+        if (it != m_vars.end()) return;
```

Comment:
As far as I understand you would ideally want something like:
```v
x = fetch_var(stream);
y = fetch_var(stream);
...
```

Unfortunately we still have to parse the string, because although the SMT-LIB2 language does guarantee the order of the returned values `((x #xval) (y #yval) ...)` to be the same as in the query `(get-value (x y ...))`, the format can vary by whitespace and number presentation format, and the algorithmic complexity of solving the problem is much worse than the parsing anyway. We still need to assign some names to the variables, and I thought original names would be more helpful than e.g. x1 through x99. Another problem to consider is whether including this optimization now will hamper implementation of future features like dynamic constraints and/or dynamically enabling/disabling constraints. It can be done, however, provided that we never add nor remove variables from the set of variables using the solver.

Reply:
Most solutions required will be very algrithmically simple - benchmark the time in the solver vs parsing?

Does/can the solver cache some information between queries?  If so, some solutions may simplify to be identical if we use the same x1/x2.

Otherwise I agree better to be verbose with variable names - but perhaps something like x1_{first10charsofvarname}, etc might make it easy to both read and parse without a find, and without having to read long var names (as everything after _ will be just a comment).

Reply:
I was able to benchmark the time spent solving and parsing on a skewed example (adding factorization to a test).

The measurements are further distorted by the fact that solver time includes IPC, while parsing is done on already buffered response (including IPC makes them comparable, but it also includes the solver serializing the response), but they show that optimizing this particular part is not going to bring spectacular performance improvements.

My patch used to generate the timing measurements:
<details>

```diff
diff --git a/include/verilated_random.cpp b/include/verilated_random.cpp
index 4eec944db..8f7469bde 100644
--- a/include/verilated_random.cpp
+++ b/include/verilated_random.cpp
@@ -364,7 +364,9 @@ bool VlRandomizer::next(VlRNG& rngr) {
 
 bool VlRandomizer::parseSolution(std::iostream& f) {
     std::string sat;
+    VlOs::DeltaWallTime t1{true};
     do { std::getline(f, sat); } while (sat == "");
+    double dt1 = t1.deltaTime();
 
     if (sat == "unsat") return false;
     if (sat != "sat") {
@@ -388,6 +390,7 @@ bool VlRandomizer::parseSolution(std::iostream& f) {
         return false;
     }
 
+    VlOs::DeltaWallTime t2{true};
     while (true) {
         f >> c;
         if (c == ')') break;
@@ -406,6 +409,8 @@ bool VlRandomizer::parseSolution(std::iostream& f) {
 
         it->second->set(std::move(value));
     }
+    double dt2 = t2.deltaTime();
+    printf("BENCHMARK %lg solver\n       vs %lg parsing\n", dt1, dt2);
 
     return true;
 }
diff --git a/test_regress/t/t_constraint_operators.v b/test_regress/t/t_constraint_operators.v
index 159f6cc52..a34b8ddd3 100644
--- a/test_regress/t/t_constraint_operators.v
+++ b/test_regress/t/t_constraint_operators.v
@@ -8,6 +8,8 @@ class Packet;
    rand int x;
    rand bit [31:0] b;
    rand bit [31:0] c;
+   rand bit [127:0] longname;
+   rand bit [127:0] d;
    rand bit tiny;
 
    typedef bit signed [63:0] s64;
@@ -23,6 +25,7 @@ class Packet;
    constraint cmps { x < x || x <= x || x > x || x >= x; }
    constraint cmpu { b < b || b <= b || b > b || b >= b; }
    constraint ext { s64'(x) != u64'(tiny); }
+   constraint slow { longname == 128'd762964647983; longname % d == 0; d > 1; d < longname; }
 
 endclass
 
```
</details>

The results are:
```
BENCHMARK 45.2291 solver
       vs 2.6237e-05 parsing
BENCHMARK 43.2575 solver
       vs 8.923e-06 parsing
BENCHMARK 0.0905919 solver
       vs 1.1884e-05 parsing
BENCHMARK 0.00446422 solver
       vs 1.2257e-05 parsing
BENCHMARK 0.00375427 solver
       vs 2.5666e-05 parsing
```

I also think that optimizing this structure can introduce unnecessary complexity to the new feature, and further performance improvements should be left for separate effort.

Reply:
Thanks for benchmarking the idea, I'm glad to be wrong in this case!